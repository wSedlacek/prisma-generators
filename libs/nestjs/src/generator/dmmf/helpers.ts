// TODO: support not string args
export const modelAttributeRegex = /(@@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export const parseDocumentationAttributes = (
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: 'model' | 'field'
) => {
  const attributeRegex =
    expectedAttributeKind === 'model'
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return undefined;
  }

  return attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
};
