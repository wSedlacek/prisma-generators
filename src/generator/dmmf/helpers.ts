// TODO: support not string args
export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export function parseDocumentationAttributes(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
) {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return;
  }
  const attributeArgs = attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
  return attributeArgs;
}
