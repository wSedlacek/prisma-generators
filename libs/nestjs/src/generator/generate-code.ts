import { DMMF as PrismaDMMF } from '@prisma/client/runtime/dmmf-types';
import path from 'path';
import { CompilerOptions, ModuleKind, Project, ScriptTarget } from 'ts-morph';

import { isDefined } from '../utils';
import { generateArgsTypeClassFromArgs } from './args-class';
import {
  argsFolderName,
  crudResolversFolderName,
  enumsFolderName,
  inputsFolderName,
  modelsFolderName,
  outputsFolderName,
  relationsResolversFolderName,
  resolversFolderName,
} from './config';
import { DmmfDocument } from './dmmf/dmmf-document';
import { DMMF } from './dmmf/types';
import { generateEnumFromDef } from './enum';
import { noop } from './helpers';
import {
  generateArgsBarrelFile,
  generateEnumsBarrelFile,
  generateIndexFile,
  generateInputsBarrelFile,
  generateModelsBarrelFile,
  generateOutputsBarrelFile,
  generateResolversBarrelFile,
} from './imports';
import { generateObjectTypeClassFromModel } from './model-type-class';
import { GenerateCodeOptions } from './options';
import { generateCrudResolverClassFromMapping } from './resolvers/full-crud';
import { generateRelationsResolverClassesFromModel } from './resolvers/relations';
import { generateActionResolverClass } from './resolvers/separate-action';
import {
  generateInputTypeClassFromType,
  generateOutputTypeClassFromType,
} from './type-class';

const baseCompilerOptions: CompilerOptions = {
  target: ScriptTarget.ES2019,
  module: ModuleKind.CommonJS,
  emitDecoratorMetadata: true,
  experimentalDecorators: true,
};

export const generateCode = async (
  dmmf: PrismaDMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop
) => {
  const baseDirPath = options.outputDirPath;
  const emitTranspiledCode =
    options.emitTranspiledCode ??
    options.outputDirPath.includes('node_modules');
  const project = new Project({
    compilerOptions: {
      ...baseCompilerOptions,
      ...(emitTranspiledCode ? { declaration: true } : {}),
    },
  });
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map((model) => model.name);

  log('Transforming dmmfDocument...');
  const dmmfDocument = new DmmfDocument(dmmf, options);

  log('Generating enums...');
  const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
    (enumDef) => enumDef.typeName
  );
  dmmfDocument.datamodel.enums.forEach((enumDef) =>
    generateEnumFromDef(project, baseDirPath, enumDef)
  );
  dmmfDocument.schema.enums
    // skip enums from datamodel
    .filter((enumDef) => !datamodelEnumNames.includes(enumDef.typeName))
    .forEach((enumDef) => generateEnumFromDef(project, baseDirPath, enumDef));
  const emittedEnumNames = [
    ...new Set([
      ...dmmfDocument.schema.enums.map((it) => it.typeName),
      ...dmmfDocument.datamodel.enums.map((it) => it.typeName),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);

  log('Generating models...');
  dmmfDocument.datamodel.models.forEach((model) =>
    generateObjectTypeClassFromModel(project, baseDirPath, model, dmmfDocument)
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmfDocument.datamodel.models.map((it) => it.typeName)
  );

  log('Generating output types...');
  const rootTypes = dmmfDocument.schema.outputTypes.filter((type) =>
    ['Query', 'Mutation'].includes(type.name)
  );
  const outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
    // skip generating models and root resolvers
    (type) => !modelNames.includes(type.name) && !rootTypes.includes(type)
  );
  const outputTypesFieldsArgsToGenerate = outputTypesToGenerate
    .map((it) => it.fields)
    .reduce((a, b) => a.concat(b), [])
    .filter((it): it is Required<DMMF.OutputSchemaField> => !!it.argsTypeName);
  outputTypesToGenerate.forEach((type) =>
    generateOutputTypeClassFromType(
      project,
      resolversDirPath,
      type,
      dmmfDocument
    )
  );
  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map((it) => it.typeName),
    outputTypesFieldsArgsToGenerate.length > 0
  );

  if (outputTypesFieldsArgsToGenerate.length > 0) {
    log('Generating output types args...');
    outputTypesFieldsArgsToGenerate.forEach((field) => {
      generateArgsTypeClassFromArgs(
        project,
        path.resolve(resolversDirPath, outputsFolderName),
        field.args,
        field.argsTypeName,
        dmmfDocument,
        2
      );
    });

    const outputsArgsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        argsFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateArgsBarrelFile(
      outputsArgsBarrelExportSourceFile,
      outputTypesFieldsArgsToGenerate
        .map((it) => it.argsTypeName)
        .filter(isDefined)
    );
  }

  log('Generating input types...');

  dmmfDocument.schema.inputTypes.forEach((type) =>
    generateInputTypeClassFromType(
      project,
      resolversDirPath,
      type,
      dmmfDocument
    )
  );

  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmfDocument.schema.inputTypes.map((it) => it.typeName)
  );

  if (dmmfDocument.relationModels.length > 0) {
    log('Generating relation resolvers...');

    dmmfDocument.relationModels.forEach((relationModel) =>
      generateRelationsResolverClassesFromModel(
        project,
        baseDirPath,
        dmmfDocument,
        relationModel
      )
    );

    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateResolversBarrelFile(
      'relations',
      relationResolversBarrelExportSourceFile,
      dmmfDocument.relationModels.map((relationModel) => ({
        resolverName: relationModel.resolverName,
        modelName: relationModel.model.typeName,
        hasSomeArgs: relationModel.relationFields.some(
          (field) => field.argsTypeName !== undefined
        ),
      }))
    );

    log('Generating relation resolver args...');
    dmmfDocument.relationModels.forEach((relationModelData) => {
      const resolverDirPath = path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        relationModelData.model.typeName
      );

      relationModelData.relationFields
        .filter((field) => field.argsTypeName)
        .forEach((field) => {
          generateArgsTypeClassFromArgs(
            project,
            resolverDirPath,
            field.outputTypeField.args,
            field.argsTypeName as string,
            dmmfDocument
          );
        });

      const argTypeNames = relationModelData.relationFields
        .filter((it) => it.argsTypeName !== undefined)
        .map((it) => it.argsTypeName)
        .filter(isDefined);

      if (argTypeNames.length) {
        const barrelExportSourceFile = project.createSourceFile(
          path.resolve(resolverDirPath, argsFolderName, 'index.ts'),
          undefined,
          { overwrite: true }
        );
        generateArgsBarrelFile(barrelExportSourceFile, argTypeNames);
      }
    });
  }

  log('Generating crud resolvers...');
  dmmfDocument.mappings.forEach((mapping) => {
    const model = dmmfDocument.datamodel.models.find(
      (it) => it.name === mapping.model
    ) as DMMF.Model;
    generateCrudResolverClassFromMapping(
      project,
      baseDirPath,
      mapping,
      model,
      dmmfDocument
    );
    mapping.actions.forEach((action) => {
      const actionModel = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;
      generateActionResolverClass(
        project,
        baseDirPath,
        actionModel,
        action,
        mapping,
        dmmfDocument
      );
    });
  });
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateResolversBarrelFile(
    'crud',
    crudResolversBarrelExportSourceFile,
    dmmfDocument.mappings.map((mapping) => {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;

      return {
        modelName: model.typeName,
        resolverName: mapping.resolverName,
        actionResolverNames: mapping.actions.map((it) => it.actionResolverName),
        hasSomeArgs: mapping.actions.some(
          (action) => action.argsTypeName !== undefined
        ),
      };
    })
  );

  log('Generating crud resolvers args...');
  dmmfDocument.mappings.forEach((mapping) => {
    const actionsWithArgs = mapping.actions.filter(
      (it) => it.argsTypeName !== undefined
    );

    if (actionsWithArgs.length) {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;
      const resolverDirPath = path.resolve(
        baseDirPath,
        resolversFolderName,
        crudResolversFolderName,
        model.typeName
      );
      actionsWithArgs.forEach((action) => {
        generateArgsTypeClassFromArgs(
          project,
          resolverDirPath,
          action.method.args,
          action.argsTypeName as string,
          dmmfDocument
        );
      });
      const barrelExportSourceFile = project.createSourceFile(
        path.resolve(resolverDirPath, argsFolderName, 'index.ts'),
        undefined,
        { overwrite: true }
      );
      generateArgsBarrelFile(
        barrelExportSourceFile,
        actionsWithArgs.map((it) => it.argsTypeName).filter(isDefined)
      );
    }
  });

  log('Generating index file');
  const indexSourceFile = project.createSourceFile(
    `${baseDirPath}/index.ts`,
    undefined,
    { overwrite: true }
  );
  generateIndexFile(indexSourceFile, dmmfDocument.relationModels.length > 0);

  log('Emitting generated code files');
  if (emitTranspiledCode) {
    await project.emit();
  } else {
    for (const file of project.getSourceFiles()) {
      file.formatText({ indentSize: 2 });
    }
    await project.save();
  }
};
