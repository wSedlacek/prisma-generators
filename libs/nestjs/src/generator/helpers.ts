import { DmmfDocument } from './dmmf/dmmf-document';
import { fieldAttributeRegex, modelAttributeRegex } from './dmmf/helpers';
import { DMMF } from './dmmf/types';

export const noop = () => {};

export const getFieldTSType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  isInputType: boolean,
  modelName?: string,
  typeName?: string
) => {
  let TSType: string;
  if (typeInfo.kind === 'scalar') {
    TSType = mapScalarToTSType(typeInfo.type, isInputType);
  } else if (typeInfo.kind === 'object') {
    TSType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else if (typeInfo.kind === 'enum') {
    TSType = `keyof typeof ${typeInfo.type}`;
  } else {
    throw new Error(`Unsupported field type kind: ${typeInfo.kind}`);
  }
  if (typeInfo.isList) {
    if (TSType.includes(' ')) {
      TSType = `Array<${TSType}>`;
    } else {
      TSType += '[]';
    }
  }
  if (!typeInfo.isRequired) {
    // FIXME: use properly null for output and undefined for input
    // TSType += ' | undefined';
    TSType += ' | null | undefined';
  }

  return TSType;
};

export const mapScalarToTSType = (scalar: string, isInputType: boolean) => {
  switch (scalar) {
    case 'ID':
    case 'UUID': {
      return 'string';
    }
    case 'String': {
      return 'string';
    }
    case 'Boolean': {
      return 'boolean';
    }
    case 'DateTime': {
      return 'Date';
    }
    case 'Int':
    case 'Float': {
      return 'number';
    }
    case 'Json':
      return isInputType ? 'InputJsonValue' : 'JsonValue';
    default:
      throw new Error(`Unrecognized scalar type: ${scalar}`);
  }
};

export const getGraphQLType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  modelName?: string,
  typeName?: string
) => {
  let GraphQLType: string;
  if (typeInfo.kind === 'scalar') {
    GraphQLType = mapScalarToType(typeInfo.type);
  } else if (typeInfo.kind === 'object') {
    GraphQLType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else {
    GraphQLType = typeInfo.type;
  }
  if (typeInfo.isList) {
    GraphQLType = `[${GraphQLType}]`;
  }

  return GraphQLType;
};

export const mapScalarToType = (scalar: string) => {
  switch (scalar) {
    case 'DateTime': {
      return 'Date';
    }
    // TODO: use proper uuid graphql scalar
    case 'UUID': {
      return 'String';
    }
    case 'Boolean':
    case 'String': {
      return scalar;
    }
    case 'ID':
    case 'Int':
    case 'Float': {
      return `${scalar}`;
    }
    case 'Json': {
      return `GraphQLJSON`;
    }
    default: {
      throw new Error(`Unrecognized scalar type: ${scalar}`);
    }
  }
};

export const camelCase = (str: string) => {
  return str[0].toLowerCase() + str.slice(1);
};

export const pascalCase = (str: string): string => {
  return str[0].toUpperCase() + str.slice(1);
};

const getInputKeywordPhrasePosition = (inputTypeName: string) => {
  const inputParseResult = [
    'Create',
    'OrderBy',
    'Update',
    'Upsert',
    'ScalarWhere',
    'Where',
    'ListRelationFilter',
    'RelationFilter',
    'Filter',
  ]
    .map((inputKeyword) => inputTypeName.search(inputKeyword))
    .filter((position) => position >= 0);

  if (inputParseResult.length === 0) {
    return undefined;
  }

  return inputParseResult[0];
};

export const getModelNameFromInputType = (inputTypeName: string) => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(inputTypeName);
  if (!keywordPhrasePosition) {
    return undefined;
  }

  return inputTypeName.slice(0, keywordPhrasePosition);
};

export const getInputTypeName = (
  originalInputName: string,
  dmmfDocument: DmmfDocument
): string => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(
    originalInputName
  );
  if (!keywordPhrasePosition) {
    return originalInputName;
  }

  const modelName = originalInputName.slice(0, keywordPhrasePosition);
  const typeNameRest = originalInputName.slice(keywordPhrasePosition);
  const modelTypeName = dmmfDocument.getModelTypeName(modelName);
  if (!modelTypeName) {
    return originalInputName;
  }

  return `${modelTypeName}${typeNameRest}`;
};

export const cleanDocsString = (
  documentation: string | undefined
): string | undefined => {
  if (!documentation) {
    return undefined;
  }
  let cleanedDocs = documentation;
  cleanedDocs = cleanedDocs.replace(modelAttributeRegex, '');
  cleanedDocs = cleanedDocs.replace(fieldAttributeRegex, '');
  cleanedDocs = cleanedDocs.split('"').join('\\"');
  cleanedDocs = cleanedDocs.split('\r').join('');
  cleanedDocs = cleanedDocs.split('\\r').join('');
  cleanedDocs = cleanedDocs.split('\n').join('');
  cleanedDocs = cleanedDocs.split('\\n').join('');

  return cleanedDocs;
};

export const toUnixPath = (maybeWindowsPath: string) => {
  return maybeWindowsPath.split('\\').join('/');
};
