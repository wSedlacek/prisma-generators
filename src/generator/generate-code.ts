import { DMMF as PrismaDMMF } from "@prisma/client/runtime/dmmf-types";
import { Project } from "ts-morph";
import path from "path";

import { noop } from "./helpers";
import generateEnumFromDef from "./enum";
import generateObjectTypeClassFromModel from "./model-type-class";
import generateRelationsResolverClassesFromModel from "./resolvers/relations";
import {
  generateOutputTypeClassFromType,
  generateInputTypeClassFromType,
} from "./type-class";
import generateCrudResolverClassFromMapping from "./resolvers/full-crud";
import {
  resolversFolderName,
  relationsResolversFolderName,
  crudResolversFolderName,
  inputsFolderName,
  outputsFolderName,
  enumsFolderName,
  modelsFolderName,
  argsFolderName,
} from "./config";
import {
  generateResolversBarrelFile,
  generateInputsBarrelFile,
  generateOutputsBarrelFile,
  generateIndexFile,
  generateModelsBarrelFile,
  generateEnumsBarrelFile,
  generateArgsBarrelFile,
} from "./imports";
import saveSourceFile from "../utils/saveSourceFile";
import { GenerateCodeOptions } from "./options";
import { DmmfDocument } from "./dmmf/dmmf-document";

export default async function generateCode(
  dmmf: PrismaDMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop,
) {
  const baseDirPath = options.outputDirPath;
  const project = new Project();
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map(model => model.name);

  log("Transforming dmmfDocument...");
  const dmmfDocument = new DmmfDocument(dmmf, options);

  log("Generating enums...");
  const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
    enumDef => enumDef.name,
  );
  await Promise.all(
    dmmfDocument.datamodel.enums.map(enumDef =>
      generateEnumFromDef(project, baseDirPath, enumDef),
    ),
  );
  await Promise.all(
    dmmfDocument.schema.enums
      // skip enums from datamodel
      .filter(enumDef => !datamodelEnumNames.includes(enumDef.name))
      .map(enumDef => generateEnumFromDef(project, baseDirPath, enumDef)),
  );
  const emittedEnumNames = [
    ...new Set([
      ...dmmfDocument.schema.enums.map(it => it.name),
      ...dmmfDocument.datamodel.enums.map(it => it.name),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, "index.ts"),
    undefined,
    { overwrite: true },
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);
  await saveSourceFile(enumsBarrelExportSourceFile);

  log("Generating models...");
  await Promise.all(
    dmmfDocument.datamodel.models.map(model =>
      generateObjectTypeClassFromModel(
        project,
        baseDirPath,
        model,
        dmmfDocument,
        options,
      ),
    ),
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, "index.ts"),
    undefined,
    { overwrite: true },
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmfDocument.datamodel.models.map(it => it.typeName),
  );
  await saveSourceFile(modelsBarrelExportSourceFile);

  log("Generating output types...");
  const rootTypes = dmmfDocument.schema.outputTypes.filter(type =>
    ["Query", "Mutation"].includes(type.name),
  );
  const outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
    // skip generating models and root resolvers
    type => !modelNames.includes(type.name) && !rootTypes.includes(type),
  );
  await Promise.all(
    outputTypesToGenerate.map(type =>
      generateOutputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options,
      ),
    ),
  );
  const argsTypesNames = outputTypesToGenerate
    .map(it => it.fields)
    .reduce((a, b) => a.concat(b), [])
    .map(it => it.argsTypeName)
    .filter(Boolean) as string[];
  if (argsTypesNames.length > 0) {
    const outputsArgsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        argsFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateArgsBarrelFile(outputsArgsBarrelExportSourceFile, argsTypesNames);
    await saveSourceFile(outputsArgsBarrelExportSourceFile);
  }

  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map(it => it.typeName),
    argsTypesNames.length > 0,
  );
  await saveSourceFile(outputsBarrelExportSourceFile);

  log("Generating input types...");
  await Promise.all(
    dmmfDocument.schema.inputTypes.map(type =>
      generateInputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options,
      ),
    ),
  );
  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmfDocument.schema.inputTypes.map(it => it.typeName),
  );
  await saveSourceFile(inputsBarrelExportSourceFile);

  log("Generating relation resolvers...");
  const relationResolversData = await Promise.all(
    dmmfDocument.datamodel.models
      .filter(model => model.fields.some(field => field.relationName))
      .map(model => {
        const outputType = dmmfDocument.schema.outputTypes.find(
          type => type.name === model.name,
        )!;
        const mapping = dmmfDocument.mappings.find(
          it => it.model === model.name,
        )!;
        return generateRelationsResolverClassesFromModel(
          project,
          baseDirPath,
          model,
          mapping,
          outputType,
          dmmfDocument,
        );
      }),
  );
  if (relationResolversData.length > 0) {
    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateResolversBarrelFile(
      "relations",
      relationResolversBarrelExportSourceFile,
      relationResolversData,
    );
    await saveSourceFile(relationResolversBarrelExportSourceFile);
  }

  log("Generating crud resolvers...");
  const crudResolversData = await Promise.all(
    dmmfDocument.mappings.map(mapping => {
      const model = dmmfDocument.datamodel.models.find(
        model => model.name === mapping.model,
      )!;
      return generateCrudResolverClassFromMapping(
        project,
        baseDirPath,
        mapping,
        model,
        rootTypes,
        modelNames,
        options,
        dmmfDocument,
      );
    }),
  );
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateResolversBarrelFile(
    "crud",
    crudResolversBarrelExportSourceFile,
    crudResolversData,
  );
  await saveSourceFile(crudResolversBarrelExportSourceFile);

  log("Generating index file");
  const indexSourceFile = project.createSourceFile(
    baseDirPath + "/index.ts",
    undefined,
    { overwrite: true },
  );
  generateIndexFile(indexSourceFile, relationResolversData.length > 0);
  await saveSourceFile(indexSourceFile);
}
