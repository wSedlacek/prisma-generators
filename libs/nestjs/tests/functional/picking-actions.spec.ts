import { NestApplication } from '@nestjs/core';
import {
  GraphQLSchemaBuilderModule,
  GraphQLSchemaFactory,
} from '@nestjs/graphql';
import { Test } from '@nestjs/testing';
import { promises as fs } from 'fs';
import { GraphQLSchema, printSchema } from 'graphql';

import { generateArtifactsDirPath } from '../helpers/artifacts-dir';
import { generateCodeFromSchema } from '../helpers/generate-code';
import { prisma } from '../helpers/prisma-template';

describe('picking prisma actions', () => {
  let app: NestApplication;
  let graphQLSchemaSDL: GraphQLSchema;

  beforeAll(async () => {
    const outputDirPath = generateArtifactsDirPath(
      'functional-picking-actions'
    );
    await fs.mkdir(outputDirPath, { recursive: true });
    const prismaSchema = prisma`
      model User {
        intIdField          Int     @id @default(autoincrement())
        uniqueStringField   String  @unique
        optionalStringField String?
        dateField           DateTime
      }
    `;
    await generateCodeFromSchema(prismaSchema, { outputDirPath });

    const moduleFixture = await Test.createTestingModule({
      imports: [GraphQLSchemaBuilderModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();

    const gqlSchemaFactory = app.get(GraphQLSchemaFactory);

    const { CreateUserResolver, FindManyUserResolver } = require(outputDirPath);
    graphQLSchemaSDL = await gqlSchemaFactory.create(
      [CreateUserResolver, FindManyUserResolver],
      { skipCheck: true }
    );

    await app.close();
  });

  it('should expose in GraphQL schema only actions chosen by single resolvers', () => {
    expect(printSchema(graphQLSchemaSDL)).toMatchSnapshot('graphQLSchemaSDL');
  });
});
