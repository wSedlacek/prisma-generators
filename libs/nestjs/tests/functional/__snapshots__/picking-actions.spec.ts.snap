// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`picking prisma actions should expose in GraphQL schema only actions chosen by single resolvers: graphQLSchemaSDL 1`] = `
"\\"\\"\\"
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
\\"\\"\\"
scalar DateTime

input DateTimeFilter {
  equals: DateTime
  in: [DateTime!]
  notIn: [DateTime!]
  lt: DateTime
  lte: DateTime
  gt: DateTime
  gte: DateTime
  not: NestedDateTimeFilter
}

input IntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

type Mutation {
  createUser(data: UserCreateInput!): User!
}

input NestedDateTimeFilter {
  equals: DateTime
  in: [DateTime!]
  notIn: [DateTime!]
  lt: DateTime
  lte: DateTime
  gt: DateTime
  gte: DateTime
  not: NestedDateTimeFilter
}

input NestedIntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

input NestedStringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringFilter
}

input NestedStringNullableFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringNullableFilter
}

type Query {
  users(where: UserWhereInput, orderBy: [UserOrderByInput!], cursor: UserWhereUniqueInput, take: Int = 20, skip: Int, distinct: [UserDistinctFieldEnum!]): [User!]!
}

enum SortOrder {
  asc
  desc
}

input StringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringFilter
}

input StringNullableFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringNullableFilter
}

type User {
  intIdField: Int!
  uniqueStringField: String!
  optionalStringField: String
  dateField: DateTime!
}

input UserCreateInput {
  uniqueStringField: String!
  optionalStringField: String
  dateField: DateTime!
}

enum UserDistinctFieldEnum {
  intIdField
  uniqueStringField
  optionalStringField
  dateField
}

input UserOrderByInput {
  intIdField: SortOrder
  uniqueStringField: SortOrder
  optionalStringField: SortOrder
  dateField: SortOrder
}

input UserWhereInput {
  AND: [UserWhereInput!]
  OR: [UserWhereInput!]
  NOT: [UserWhereInput!]
  intIdField: IntFilter
  uniqueStringField: StringFilter
  optionalStringField: StringNullableFilter
  dateField: DateTimeFilter
}

input UserWhereUniqueInput {
  intIdField: Int
  uniqueStringField: String
}
"
`;
